serial_protocol.md

```diff
@@ -0,0 +1,186 @@
+###Usage
+	SET $PATH $DATA
+###Response
+	+OK
+	-FAIL
+###Examples
+	>>SET /user/aturing/first Alan
+	<<+OK
+##SET$
+Similar to SET above but used to write multiline strings or raw binary data. Data format is similar to the batch string ($) return type, we specify the $DATA_BYTE_COUNT on the same line as SET$ then a newline and all data. However which the batch string ($) return type returns data json escaped and quoted you may provide raw data and we will handle the escaping.
+
+Receiver will wait until a timeout for client to send $DATA_BYTE_COUNT worth of data before becoming responsive again.
+###Usage
+	SET$ $PATH $DATA_BYTE_COUNT
+	$DATA
+###Response
+	+OK
+	-FAIL
+	-FAIL_TIMEOUT
+###Examples
+	>>SET /user/aturing/address 24
+	>>78 High Street,
+	>>Hampton 
+	<<+OK
+	
+##SET{+,*,#,.,?}
+Same as SET but will force the value to be stored in the given type or return an error if we cannot parse it as that type.
+###Usage
+	SET+ $PATH $VALUE
+	SET: $PATH $VALUE
+	SET? $PATH $VALUE
+###Response
+	+OK
+	-INCORRECT_TYPE
+###Examples
+	>>SET? /user/aturing/was_human true
+	<<+OK
+	>>SET? /user/aturing/was_human He was not a computer.
+	<<-INCORRECT_TYPE
+
+##REMOVE
+Deletes the value located at the path provided.
+###Usage
+	REMOVE $PATH
+###Response
+	+OK
+	-FAIL
+###Examples
+	>>REMOVE /user/aturing
+	<<+OK
+
+##PUSH
+Adds a value to the list located at the path provided and returns the key at which the new object is located.
+###Usage
+	PUSH $PATH $DATA
+###Response
+	$KEY
+###Examples
+	>>PUSH /user/aturing/login_timestamps 1455052043
+	<<+-K94eLnB0rAAvfkh_WC2
+
+##PUSH$
+Similar to PUSH but used to write multiline strings or raw binary data. Data format is similar to the batch string ($) return type, we specify the $DATA_BYTE_COUNT on the same line as SET$ then a newline and all data. However you are not required to json escape all of your data, that will be handled for you.
+
+Receiver will wait until a timeout for client to send $DATA_BYTE_COUNT worth of data before becoming responsive again.
+###Usage
+	PUSH$ $PATH $DATA_BYTE_COUNT
+	$DATA
+###Response
+	$KEY
+###Examples
+	>>PUSH /user/aturing/quotes 91
+	>>We can only see a short distance ahead,
+	>>but we can see plenty there that needs to be done.
+	<<+-K94eLnB0rAAvfkh_WC3
+
+##BEGIN_STREAM
+Used to register to receive a stream of events that occur to the object at the provided path.
+
+After registering you will start receiving events on the response line. They will be formatted as one line with the event type {PUT,PATCH,etc..} followed by the sub_path that changed and the other line with the data associated with that event type. This data will be formatted similar to GET results and can have multi-line batch strings (*) or json strings (&).
+
+The event stream will continue until you send CANCEL_STREAM.
+
+When there is an ongoing event stream no other commands can be processed until you call END_STREAM as the event stream owns the return line. 
```

Comment:
re: "no other commands can be processed", since our protocol is serial based the client will have to poll `if Serial.available() Serial.readline()` for stream result, maybe it wouldn't hurt to make that more explicit with a command? that way this could be multiplexed with other operation.